src/components/container/InsightMatrix/InsightMatrix.tsx

```diff
@@ -0,0 +1,146 @@
+    const startDate = add(today, { days: -daysToCompute });
+
+    const getDataProvider = <T extends InsightMatrixDataConfiguration>(configuration: T): InsightMatrixDataProvider<T> => {
+        if (isSurveyDataType(configuration.rawDataType)) {
+            return { configuration: configuration, loadData: getSurveyDataProvider(configuration.rawDataType) };
+        }
+        return { configuration: configuration, loadData: getDailyDataTypeDefinition(configuration.rawDataType).dataProvider };
+    };
+
+    const getGroupByDataProvider = (): InsightMatrixDataProvider<InsightMatrixGroupByDataConfiguration> => {
+        if (props.preview) {
+            return createPreviewDataProvider(props.groupByDataConfiguration);
+        }
+        return getDataProvider(props.groupByDataConfiguration);
+    };
+
+    const getComparisonDataProviders = (): InsightMatrixDataProvider<InsightMatrixComparisonDataConfiguration>[] => {
+        if (props.preview) {
+            return props.comparisonDataConfigurations.map(createPreviewDataProvider);
+        }
+        return props.comparisonDataConfigurations.map(getDataProvider);
+    };
+
+    useInitializeView(() => {
+        const groupByDataProvider = getGroupByDataProvider();
+        groupByDataProvider.loadData(startDate, today).then(result => {
+            return { configuration: groupByDataProvider.configuration, result: result };
+        }).then(setGroupByData);
+
+        const comparisonDataProviders = getComparisonDataProviders();
+        Promise.all(comparisonDataProviders.map(comparisonDataProvider => comparisonDataProvider.loadData(startDate, today).then(result => {
+            return { configuration: comparisonDataProvider.configuration, result: result };
+        }))).then(setComparisonData);
+    }, [], [props.preview, props.groupByDataConfiguration, props.comparisonDataConfigurations]);
+
+    if (!groupByData || !comparisonData) {
+        return <div className="mdhui-insight-matrix" ref={props.innerRef}><LoadingIndicator /></div>;
+    }
+
+    const thresholdDaysLookup = computeThresholdDays(startDate, today, groupByData);
+
+    const allThresholds = groupByData.configuration.thresholds.map(threshold => threshold.label).concat(NotEnteredThreshold);
+    const filteredThresholds = allThresholds.filter(threshold => Object.keys(thresholdDaysLookup).includes(threshold));
+
+    const getDefaultDataLabel = (configuration: InsightMatrixDataConfiguration): string => {
+        if (isSurveyDataType(configuration.rawDataType)) {
+            return configuration.rawDataType.stepIdentifier;
+        }
+        const labelKey = getDailyDataTypeDefinition(configuration.rawDataType).labelKey;
+        return labelKey ? language(labelKey) : configuration.rawDataType;
+    };
+
+    return <div className="mdhui-insight-matrix" style={{ gridTemplateColumns: '112px 1fr ' + Array(Math.max(0, comparisonData.length - 1)).fill('1fr').join(' ') }} ref={props.innerRef}>
```

Comment:
Is a LoadingIndicator needed?

Reply:
The loading indicator is up on line 87: https://github.com/CareEvolution/MyDataHelpsUI/pull/352/files#diff-37c4e9f5691934f943b8da5a8f6e08abbb815de42c9da6481bf3b55a0c6abe98R87